### ISSUE TITLE
Provide method and/or config to programmatically scroll to ListView index

### ISSUE THREAD

**Issue by ariejdl** [flutter/flutter#12319]
The ListView is successful at displaying large lists, however in some circumstances, such as a reader application it is useful to be able to focus to a certain list item offset, the iOS `UITableView` has `- (void)scrollToRowAtIndexPath:(NSIndexPath *)indexPath atScrollPosition:(UITableViewScrollPosition)scrollPosition animated:(BOOL)animated` and on Android the `ListView` has either `setSelection(index);` or `smoothScrollToPosition(index)`.  Please can you add this feature?

**Comment by ariejdl** [comment]
@Hixie is this a reasonable feature request?

**Comment by Hixie** [comment]
It's certainly a reasonable request... how easy it is to implement is a different matter. :-)

For fixed-item-height list views it's easy to work around (just jump to the index * the height). For variable-height list views, it's really difficult in the current setup (because we don't know how far to scroll). It's the same thing that's blocking https://github.com/flutter/flutter/issues/10826.

**Comment by ariejdl** [comment]
@Hixie please forgive my ignorance, but does it have to do with `endScrollOffset < scrollOffset` in `performLayout` in https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/rendering/sliver_list.dart#L213 ?

**Comment by Hixie** [comment]
That's part of the code that makes sure that widgets that aren't visible aren't laid out, yes. And that's the reason we don't know the offset for the widgets that aren't visible.

**Comment by sroddy** [comment]
I have to add that I spent quite a few time in the past to figure out the fact that this feature is missing as I consider it being a critical requirement for a lot of common scenarios (chats, readers, contact lists just to name a few...) and for sure at least 2/3 of our apps.

While I understand the complexity of an animated version of this feature, I thought that a non-animated `jumpToIndex(int index, {double position})` method shouldn't be that hard to implement, as it would require to just trigger a brand new render/layout of the visible area of the list starting from the element at the index we need to make visible (instead of index 0) and then walk up and down to fill the holes.

**Comment by sroddy** [comment]
another possible approach would be to be able to correctly estimate the height of the rows upfront and provide a correct offset to the `jumpTo` method. In one my specific use-cases the only variable part of the equation was to figure out how many rows a text label would have occupied when being rendered on the specific container. iOS has a method to achieve this (https://developer.apple.com/documentation/foundation/nsattributedstring/1529154-boundingrectwithsize), I''m wondering how to achieve the same goal with Flutter.

**Comment by GRouslan** [comment]
In case of smooth scrolling, if it is 1 element visible and 1001 is requested, no need to scroll through all 1000 elements.

**Comment by Hixie** [comment]
This is definitely something we want to offer, we just haven't implemented it yet. The general case (animating to an item selected by index) is non-trivial, but we sort of need to know what our approach for that will be before we can design the API for the simpler cases as well.

**Comment by DmitryIvanitskiy** [comment]
I made a small test app implementing ScrollToLine functionality, it should works also with mixed (different lineHeight) lists. The source: https://github.com/DmitryIvanitskiy/Animation/blob/master/main.dart
My approach is:  ScrollToLine  renders list twice: 
1. Render list truncated from start to target line (so ScrollController.position.maxScrollExtent is  offset of the last line - save it).
2. Render full list and animateTo offset we get on step 1. 
I encountered the following problem - ScrollController.position.maxScrollExtent didn't changes instantly after setState, but only after some (random, but short) time, solution made code extremely ugly... 
Is there a way to set a listener/callback directly to ScrollController.position.maxScrollExtent ?

**Comment by eseidelGoogle** [comment]
This was brought up in https://proandroiddev.com/quire-building-a-sophisticated-mobile-app-with-googles-flutter-2770065284d6 (from quire) just now.

**Comment by volgin** [comment]
I many use cases jumpTo(_index) is critical, while scrollTo(_index) is unnecessary or even undesirable.

If a user is in the middle of a long list, tapping to see the details and then coming back, scrolling all the way from the top to the desired position is a bad UX, whether it's manual or programmatic. I believe that jumpTo() is a major UX improvement in 100% of "coming back" use cases, while scrollTo is applicable only to a small subset of use cases.

**Comment by rrousselGit** [comment]
@volgin "coming back" isn't the only use-case.
You could have a button on screen which allows you to jump to a specific item. 

____ 

The key point is, `scrollTo(index)` should have the same options as ` Future<Null> 
 Scrollable.ensureVisible(BuildContext context, { double alignment: 0.0,  duration: Duration.zero, curve: Curves.ease, })`

I'll admit I expected `Scollable.ensureVisible` to be the solution.
After all, when using a `SingleChildScrollView` coupled with a `Column` instead of a `ListView`, then `Scrollable.ensureVisible` achieve pretty good results.

**Wouldn't it be possible to instead extend `Scrollable.ensureVisible` to look for a widget possessing a specific `Key` ? And then considering `scrollTo(index)` to be done by searching for a  `new ObjectKey(index)` ?**

It should be possible using `Element.visitChildElements` combined with `Element.widget.key`.

**Comment by rrousselGit** [comment]
By following this logic. If the widget possessing the `Key` we are looking for is not rendered yet ; we could lay out extra slivers until we find one matching. 

At this point we should have the  index of the sliver we want to scroll to (and potentially the distance too).

**Comment by seenickcode** [comment]
This would be great for `PageView` as well.

**Comment by marica27** [comment]
I'm implementing gallery of photos and have a button "scroll to chosen date when photo was made". So scrollToPosition also important for me.  

Calculating offset is not very accurate and on the long amount of data (2-4k photos) it's not right. And I could change the layout, so I would have to rewrite logic for calculating offset, especially to support changing orientation.

**Comment by CyanBlob** [comment]
I would also love to have this feature. As it stands, I'm not really sure of a great way to implement a chat app where the message list starts scrolled to the latest message, especially since not all rows will have the same height

**Comment by zoechi** [comment]
@CyanBlob set https://docs.flutter.io/flutter/dart-core/bool-class.html to `true` and be done.

**Comment by marica27** [comment]
@zoechi Along with datePicker feature I have a draggable scrollbar showing date next to it. Have you seen Google Photos app or in Samsung Gallery App? So I want the same. 
And now I'm realising that for this anyway I need to know the exact offset of all item. So scroll to position will not help me with it. And as i know offset for all items I can use it to scroll or jump to selected date.

**Comment by zoechi** [comment]
> I need to know the exact offset of all item. So scroll to position will not help me with it.

I don't follow. If you know the offset, then you can just scroll there like `_scrollController.animateTo(1500.0,
          duration: const Duration(milliseconds: 100),
          curve: Curves.easeOut);`

**Comment by marica27** [comment]
sorry for pure explanation. My list has 2 features. One is scroll to item with selected date and other is draggable scrollbar with label showing date for current position (as in google photos app). 
When I wrote comment here first time I did not implement yet a draggable scrollbar so I did not know I need to know anyway offset for every item. 
So, I'm going to calculate height of every item and it will resolve  2 my  features.

**Comment by CyanBlob** [comment]
@zoechi I had tried setting `reverse: true`, but that caused the newest messages to be on the top, which isn't what you want for a messaging window. As it turns out, I just wasn't thinking about the problem properly. The simple solution was to reverse the order in which I pull data from the server, along with reversing the render direction of the ListView

**Comment by marcglasberg** [comment]
@Hixie The core of this problem is that `ListView`s use a pixel offset, only. It will all be solved if you could use a `pixelOffset`+`startIndex`. At present, `startIndex` is implicitly always 0. So, if you want to start displaying item 1000, it has to calculate the size of all items from 0 to 999, just to know the pixel offset. Instead, to move to item 1000, some `navigateToItem(1000)` method could set `pixelOffset` to 0, and `startIndex` to 1000. Then, if you scroll to list to previous items you would have a negative pixelOffset. In other words, you just move the "origin" around.

**Comment by rrousselGit** [comment]
@marcglasberg `Viewport`,  a lower layer of `ListView`, is already able to have a start index different from 0.  

The biggest difficulty is what comes around such feature. Like scrollbars, animated scroll and slivers.

**Comment by rrousselGit** [comment]
And then we have things like `CustomScrollView` or `SingleChildScrollView`.

What would `jumpToIndex` do when associated to these ? And since `Scrollable` and `ScrollController` don't even have a notion of index to begin with, where would we put such method ?




Somehow I think the best solution we can have is a `ScrollController.jumpToKey(Key key)`.

**Comment by marcglasberg** [comment]
@rrousselGit This is an easy solution for the ListView. Isn't the present issue is about the ListView? ScrollController.jumpToKey(Key key) is more generic, yes, but it then needs to calculate the height of all items in between (that are not even going to be displayed) and that's not going to work for ListViews with thousands of items (and maybe these items have to be download from the network).

**Comment by rrousselGit** [comment]
@marcglasberg The problem is, the scroll ecosystem is so much more complex than that.

For example you can do the following :

```dart
NotificationListener<ScrollNotification>(
  onNotification: (scrollNotification) {
    print('''
${scrollNotification.metrics.extentBefore}
${scrollNotification.metrics.extentAfter}
''');
  },
  child: ListView.builder(
    itemBuilder: (context, index) {
      return ListTile(
        title: Text("Hello World"),
      );
    },
  ),
),
```

This simple code snippet implies that any widget can get notified of scroll events ; as long as they have a Scrollable somewhere in their children.

Such change means we'd have to go through all existing `ScrollNotification` listeners and make sure they are compatible with the new API.

**Comment by marcglasberg** [comment]
So keep the ListView class as is, and add a navigateToIndex method that needs to calculate each item size. It can be used with:
1) Lists of any size, with fixed extent items.
2) Small do medium lists, with items in memory.
3) Small lists, with items downloaded from the network (e.g. Firebase).

Then, we create another class (SkippableListView, or CursorListView, or a better name) with a navigateToIndex that uses pixelOffset+startIndex and can be used with:
4) Big lists, with items in memory.
5) Medium and big lists, with items downloaded from the network (e.g. Firebase).

A proof of concept for this second list can easily be created using a technique like this: https://stackoverflow.com/questions/44468337/how-can-i-make-a-scrollable-wrapping-view-with-flutter

**Comment by marcglasberg** [comment]
@rrousselGit What would ScrollController.jumpToKey(Key key) do to solve the problem of scrolling to the 100000th item of a ListView with a builder? You dont know the key of that item.

**Comment by marcglasberg** [comment]
@marica27  Suppose you want to scroll to index 10000. You make pixelOffset = 0, and startIndex=10000. Then, when the builder asks for the item corresponding to some index, you simply return the item of startIndex+index. 

At first, since we jumped to pixelOffset=0, the list will ask for the item at index = 0, right? But you will 
return the item at startIndex+0=10000. If that item doesnt fill the screen, the list will ask for the next one, with index=1, and you will return the item at startIndex+1=10001. And so on. 

So you see it will draw items starting at 10000. Now you just let the user scroll at will, and dont need to change the startIndex again, unless you need to do another jump. Say, that now you want to jump to item 5372. You simple make startIndex=5372 and pixelOffset=0.

Got it? The list doesnt need to check any items in-between, because it is just moving the ORIGIN. It's saying that pixelOffset=0 correspond to the item with the index you want to jump to. And then you can simply go to pixelOffset=0 to display that item.

The reason why you now need the ListView to scroll to negative values is that if pixelOffset=0 corresponds to index 5372, the user may want to scroll back to item 5372, which means a negative pixelOffset.

So, to start you need a unbounded ListView which can scroll to negative offsets. This link shows the technique to do just that:  https://stackoverflow.com/questions/44468337/how-can-i-make-a-scrollable-wrapping-view-with-flutter

I don't think it would take you more than a day to implement this. Let me know if you need me to help you with that. I see that you are also in the Flutter Study Group, and I will say hi to you there if need to contact me.

**Comment by rrousselGit** [comment]
While I can understand the need for a quick solution; I don't think we should take this approach. 

I mean we're talking about forking `ListView` (and other scrollables too eventually) for just one method. That will be really confusing for users. Especially considering it'd require a custom API, ScrollController and would most likely be incompatible with `CustomScrollView`.


I think we should take a step back and see how we can change the current API to make it compatible with such jump.


In the meantime there's an easy workaround :

```dart
    SingleChildScrollView(
      child: Column(
        children: <Widget>[
          ...
        ],
      ),
    )
```

combined with `Scrollable.ensureVisible(context)`.

**Comment by marcglasberg** [comment]
@rrousselGit We are obviously not talking about the same thing here, since `Scrollable.ensureVisible(context)` could not possibly solve the problem of having, say a database with 1 million items, and then jumping to the 100000th item without having to download all those items.

**Comment by ozexpert** [comment]
Any updates on supporting this feature? I need a solution and is a critical feature to me.

**Comment by Hixie** [comment]
There is a way to jump to a scroll offset.
If all your items are the same dimensions, then you can jump to the index * the height of your items to get to the specific item.

If your items are not all the same dimensions, then the problem is quite difficult and it's not at all clear how we should solve it. We need to know the scroll offset before we can begin layout, otherwise e.g. the scroll bar won't be able to show reasonable results. We can't find out the scroll offset without building and laying out everything between the current offset and the new one. We can't even know what direction we should be doing that in since the list could be infinite in both directions.

**Comment by ozexpert** [comment]
That is sad. Coming from Ionic framework, this feature is pretty much supported from v1.

**Comment by SteveAlexander** [comment]
I have scroll to item working on my flutter project, even though many items are variable height. The thing is, I can predict the height of each item from its data, without needing to render it. So although it's a little complicated, I can accurately calculate scroll positions.

But this would be so much easier if CustomScrollView were to work in a different way. I spent a day or so on this, but I'm afraid the relationship between ViewPort and ScrollPosition and various other collaborating objects got too complex for me in the time I had available. I'll try and describe what I was thinking below.

1. For all of my needs so far, I don't need a scroll bar. I wonder whether supporting a scroll bar can be made optional in some way?

2. In many cases we're talking about here (e.g. infinite scroll in both directions, large numbers of items of variable height or height that is only known on render) the scroll-bar dimensions are kind of ambiguous or guesses anyway. Like, if I scroll half way down a large number of variable-height items, and the bottom half of these aren't rendered yet, then what should a scroll-bar look like?

I wonder if a different kind of scroll view could be made to support this?

Starting from the idea of a ViewPort, and a particular "anchor" sliver being positioned somewhere within the viewport (e.g. top or centre), the ViewPort needs to render that sliver, and to be able to render the surrounding slivers to cover the viewport and whatever area above and below it are needed to enable smooth scrolling. Consider the scroll position as moving positively or negatively relative to the anchor sliver.

Then delegate the model behind the scrollbar to a specific object that knows how to make a scrollbar make sense for this particular case (it may be different each time), or allow to say "there is no scroll bar".

**Comment by SteveAlexander** [comment]
Thanks @rrousselGit, it makes sense that these features rely on the current way of working.

I'm not proposing changing the current scrollview to work a different way. Rather, I'm asking what if there were a different (additional) kind of scroll view could be made to support viewport-oriented / anchored working, knowing that this additional kind of scrollview would not necessarily support the other features you mentioned.

**Comment by marica27** [comment]
I've dig in scrolling api twice. Everywhere I see  that offset should be positive `assert(offset >= 0)`. But it's still complicated for me to understand it good.
First I decide to make a RenderSliverList with `initialIndex` parameter and in `performLayout` to do what it usually does if ScrollController has `initialOffset` - do `layout` for items until offset is less `initialOffset`. So I do `layout` for items until index is less `initialIndex`. But cons of this solution is it has to `layout` all 1000 items if you need to jump to 1001.  I call it `IndexedListView`.
Second solution actually was inspired by this [SO](https://stackoverflow.com/questions/44468337/how-can-i-make-a-scrollable-wrapping-view-with-flutter) and with help of @slightfoot. I call `DividedListView`. It divides list by selected index, builds stack with negative list and positive list. When user scrolls, positive scrollController changes offset of negative scrollController.  
You can find code [here](https://github.com/marica27/flutter_indexed_listview)
I notice a bug in it (negative list is seen behind when after jump), not in the video
![devided_list_view](https://user-images.githubusercontent.com/38860483/43480066-6f06a306-9502-11e8-9be0-0553e1120db5.gif)

**Comment by marica27** [comment]
I've watched [DartConf 2018 talk](https://youtu.be/4lLI-hZvGg0?t=3h50m28s) "It’s bigger on the inside: mind-bending scrolling in Flutter" of Eugenio Marletti @Takhion and my 3rd experiment would be create Viewport to support negative offset, and IndexedScrollController which can `jumpTo(index)`
I would not support other features that usual ListView has. In my app I need only `jumpTo(index)`.  I've tried calculate items height but it became too complicated and I gave it up.

It would be nice if somebody who knows scrolling api really good could make simple IndexedScrollView without support features which was mentioned above.

**Comment by sroddy** [comment]
@Hixie 
> We need to know the scroll offset before we can begin layout, otherwise e.g. the scroll bar won't be able to show reasonable results. We can't find out the scroll offset without building and laying out everything between the current offset and the new one. We can't even know what direction we should be doing that in since the list could be infinite in both directions.

This is a very complex problem if you want to solve the issue with offsets, but definitely doable, even if not trivial, if you approach the problem considering a `scrollToIndex` API and you introduce a concept of estimated offset, which is something that is stable when the list is not scrolling, but can change and adapt while rendering scrolling animations. Talking high-level:
1) you take note of the current visible indexes to determine the direction for the scrolling and rely on a provided estimate of the item size.
2) you start offstage and lay out the desired widget for the given index at the given position in the viewport.
3) you lay out enough indexes in the right direction in order to achieve the desired scrolling simulation.
4) you go onstage and render a scroll view that starts from the furthest index. the offset here, for what regards the scrollbar etc, would be synthetic and computed from a given estimate of widgets' height (or a computed average of all the now rendered widgets) multiplied by the number of items in the list.
5) you `scrollTo` the desired item offset

**Comment by rrousselGit** [comment]
>  and rely on a provided estimate of the item size.

Interesting, but what is it used for?

**Comment by sroddy** [comment]
The idea is that the scrollview will consider that to be the size of any non rendered items (and use it e.g. for the offset, the scroll bar and the scroll physics).

It's the same approach adopted by Apple with their UICollectionView:
https://developer.apple.com/documentation/uikit/uicollectionviewflowlayout/1617709-estimateditemsize

**Comment by debuggerx01** [comment]
I've writen a simple demo to show how I sovle this problem by using my widget named [RectGetter](https://pub.dartlang.org/packages/rect_getter):
![demo](https://user-gold-cdn.xitu.io/2018/8/3/164fc96b7b957602?imageslim)

source code:  [JumpToIndexDemo](https://gist.github.com/debuggerx01/b8ef756ee02b3eb82ec3702f14ba94e6)

**Comment by marcglasberg** [comment]
@marica27 Problem with your DividedListView is that the positive list is on top of the negative list, so you can't actually click anything on the negative list. To see the problem, please substitute `child: new Center(child: new Text('Card $index')),` with `child: MaterialButton(onPressed: () => print(index), child: new Text('Click me $index'))),`. You will be able to click the blue items but NOT the red.

**Comment by marcglasberg** [comment]
I have just published this:

   https://pub.dartlang.org/packages/indexed_list_view

Similar to a ListView, but lets you programmatically jump to any item, by index. Items may have different sizes, it doesn't matter, the jump is always cheap and it doesn't need to build all widgets between the old and new positions. You can jump millions of items, instantly. Please run the example in the Example tab.

Currently, the list is always infinite, both to positive and negative indexes. In other words, it can be scrolled indefinitely both to the top and to the bottom. This is useful for some use cases, but it's a serious limitation for others. I don't understand the scrolling API well enough to allow passing lower and upper bounds as to make the list finite, so I would very much appreciate if someone could help me with that.

**Comment by debasmitasarkar** [comment]
@Hixie I am building an chat app..I am getting chat messages from Firebase database..Whch can be of dynamic number,using ListView.builder for messages to display.How can I autoscroll to the end of my list and on adding of message I need to autoscroll again..How can I implement this?Is there any easy way?

**Comment by AppleEducate** [comment]
https://stackoverflow.com/questions/46864763/how-to-automatically-scroll-to-end-of-grid-view

**Comment by debasmitasarkar** [comment]
@AppleEducate it helps when I am sending a message.But when I am getting all the messages for the first time , I mean opening Chat app  and getting previous messages, I need something to autoscroll at the end of the current listview.

**Comment by zoechi** [comment]
@debasmitasarkar  you might want `reverse: true`. With this you get that behavior automatically.

**Comment by debasmitasarkar** [comment]
@zoechi Yes I can do that..But my chat messages are sorted in a timely manner.So it would reverse the logic and show old data at the end and newest at the top

**Comment by zoechi** [comment]
Sorting the data in reverse order shouldn't cause any difficulty.
Updating the scroll position will be much less reliable.

**Comment by debasmitasarkar** [comment]
@zoechi Thanks Friend!! I actually solved the issue with your tips 👍

**Comment by ozexpert** [comment]
When would we have a feature like scrollToIndex()? Any estimates on when this would be fixed?

**Comment by rockingdice** [comment]
The index is presenting an abstract thing without pixels which could lead to the calculation of the height, so if you want to get a tween animation from one index to another, the calculation is not avoidable.

For the best result, we all want a scrollToIndex(), so the ListView(not ScrollView, which doesn't handle indices) should take the responsibility to calculate the offset for the tween animation. 

The height of a row (widget) can be calculated after the layout is finished. Then the height information is able to be used to do the scrolling. 

Here come two ways to do it:
1. Because there's a tween animation, the calculation could be done during the animation. The animation stops when the desired index is reached. (@debuggerx01 's implementation)
2. Precalculate the offset to the new index. This cannot be done until all the widgets that would be visible during the animation are finished with layouts.

The second is better since you know all the information ahead of time, so you can do better curve animations to present to users.

But it's easier to do jumpToIndex().

Follow current design, I think you can change the absolute pixel offset to a relative one, which is relative to an arbitrary start index.

The way is simple: 
After jumpToIndex(), it will change the starting index to a new value and the offset will be reset to 0.0. Then scrolling offset is negative or positive if you scroll up or down. No calculations are needed.

I agree jumpToIndex() will be much better UX than scollToIndex if the range is very long. On the opposite, scrollToIndex() is better for short range with an animation.

scrollToIndex() could lead to serious performance issues, which is normal and necessary because the calculation cannot be avoided.
The user should take the responsibility to decide whether they use jumpToIndex() or scrollToIndex().

But currently, I think jumpToIndex is fundamental to all users and easier to achieve, so I would request it as my first priority.

**Comment by rrousselGit** [comment]
I don't think the difficulty is in `ListView`, but in how to make it compatible with `CustomScrollView`.

We have some examples earlier about how to implements this with ListView relatively easily. 
The thing is, `CustomScrollView` is a must too to truly say we solved this problem.

The main reason being things like `SliverAppBar`. This kind of appbar is relatively common, and we definitely want to support this in combination with the `jumpToIndex`.

**Comment by RonenR** [comment]
The functionality of jumpToIndex(index) would be really helpful! 

If it helps anyone - I can share my idea for a workaround, which I implement for a reader application:
Persist the offset, so I can jumpTo(offset) with the controller next time. Of course, it doesn't replace the need for jumpToIndex, but it's something...

**Comment by tje3d** [comment]
Just developing a chat app, and really miss jumpToIndex/Item and KeepScrollOffset when add item to top of list.

**Comment by zoechi** [comment]
@tje3d just use `ListView(reverse: true, ...)`

**Comment by pedromorgan** [comment]
Was expecting this to be a "standard feature".. as per original comment, and a few months later.........

In my case, I got a list of 580 items, pre loaded as fundamental part of app/global...

I just want to scrollTo/focusTo an item.. or am i missing something...

**Comment by pedromorgan** [comment]
@zoechi sigh!.. https://pub.dartlang.org/packages/indexed_list_view

Its a bit daft IMHO that this issue is even here.... this is a showstopper for my app and experimenting with workarounds.. as a newbie (not an issue on other platforms/frameworks.. sigh)  !!

**Comment by pedromorgan** [comment]
Related is #12319

**Comment by rajeshzmoke** [comment]
i have a ListView with a Column of dynamic widgets and those dynamic widgets can add more dynamic widgets..i'm confused on how to scroll up or down to that specific widget when it is added.

**Comment by wtoalabi** [comment]
Hello @eseidelGoogle @Hixie @maguinis @zoechi @rrousselGit @ozexpert @SteveAlexander 
So, it is 2019...over two years since this issue was raised and still, there is no solution...or is it that I dont know what to ask/look for?
While I love the work-around by the community, for instance, the indexed_list_view package, unfortunately, it does not sole my use-case.
I dont want infinite scrolling.
I am building a bible app, so, the verses are loaded within the Chapter list.
I know them.
I know the count within the list, even before the widget builds.
So, if I want to jump to verse 20 (index 19) within the list, how can I make that happen?

It will be disingenuous to ask users to start from the start of the list every time, whenever they are looking for a verse in the middle.
And no, I cant have the same height for all the verses.
Some verses span 8 lines, some are just one line only.
This will look so bad from UI/UX point of view.
The earlier mentioned work-around seemed to work, it locates the exact index, but then it creates infinite scrolling before and after the list items.
The reason for this is that the package caters for infinte scrolling...and the author @marcglasberg has no ETA for an alternative: https://github.com/marcglasberg/indexed_list_view/issues/8

The other alternative suggested is SingleChildScrollView which sounds great on paper, but it cannot also work for my use-case.

For instance, I need to apply a globalKey to each item in the list BEFORE rendering and somehow call the required key with: `Scrollable.ensureVisible(_key);`

I declared: `Map<dynamic, GlobalKey> _key = {};`
at the top of the class and then 
```
model.currentChapter.forEach((each){
          _key[each.verseID] = GlobalKey();
        });
```
But when I asked for `print("${_key[15].currentWidget}");`
It returns null.
It seems to be an overkill anyways.
So, I really need your help to get this off the ground.
This feature is literary the only thing remaining for the bible reading part of the app to be completed.

Thanks everyone!!!

**Comment by jerrywell** [comment]
HI, we are Quire and just published the library https://github.com/quire-io/scroll-to-index. you can use it to scroll to the index with fixed or variable row height. Our app use this mechanism in many buit-in scrollable widget from Flutter. so it may fits most case of your usage. Enjoy it :)

**Comment by wtoalabi** [comment]
@jerrywell That package looks so detailed and well thought-out.
I am more than happy to take it for a spin and report my feedback!

Thanks.

**Comment by wtoalabi** [comment]
Well I must way, I am really impressed!
It worked like a charm.
Especially the highlight part.
Wonderful!

If I must add anything though, then it would be the example part.
It was a bit confusing to know which part is the index (as in the index in the array) and which index is the real data you are playing with.

Not that it would be too hard to figure out eventually, I mean, it took me, a total noob, just about 15 minutes.

But it would be much better if we could make it a bit simpler from the get-go.
I am happy to supply a PR that imitates a real world data in the example.

But overall...this is it....this is what I have been looking for!

Thanks!

NB: And cool app by the way!!!

**Comment by valerybodak** [comment]
@jerrywell 
Hi. Can we use your library (https://github.com/quire-io/scroll-to-index) for ListView with ExpansionTile items with unknown height?
Thanks in advance.

**Comment by jerrywell** [comment]
Sorry, i never used ExpansionTile before. but it's a widget based class. so you can wrap it with AutoScrollTag(). in this official [example](https://flutter.dev/docs/catalog/samples/expansion-tile-sample), just replace EntryItem(data[index] with:

```
AutoScrollTag(
  child: EntryItem(data[index]
)
```
and add a AutoScrollController into ListView.

you can also combine scroll-to-index's [example](https://github.com/quire-io/scroll-to-index/blob/master/example/lib/main.dart) and ExpansionTile example to see if it works. 
Basically, any class extending Flutter widget should work as expected :)

**Comment by valerybodak** [comment]
@jerrywell 
Hi, your library works fine with ListView. But it will be very helpful if you will implement scroll to position for **SingleChildScrollView** widget also (https://api.flutter.dev/flutter/widgets/SingleChildScrollView-class.html).

**Comment by wtoalabi** [comment]
Maybe he can...but should he?
I tried using SingleChildScrollView widget for the exact same problem this package solved but it didnt work out...because the use-cases are totally different.

Aside from performance concerns...which is a big factor when considering a fairly large list, SingleChildScrollView works with a list you have to manually input into the `<Widget>[]` list.

That way, you can manually input the key into each of them and then call each via its key through: 
`Scrollable.ensureVisible(_key);`

So, if you already have access to the keys of each or any of the item you need within the list, I wonder why you would need @jerrywell 's library to do that?

**Comment by oseiasmribeiro** [comment]
Wtoalabi could you give me an example of how you used scroll-to-index with the Bible? I want to do something as the image illustrates:
![54072966-4dbc7200-4260-11e9-9973-f8ed572a71f3](https://user-images.githubusercontent.com/36827173/56771044-f2046480-678b-11e9-8514-673f03f8a464.jpg)

**Comment by wtoalabi** [comment]
@oseiasmribeiro gladly!
Here is a gist I created with my code samples and brief descriptions of what I did.
Take a look and maybe discuss further implementation details.
https://gist.github.com/wtoalabi/51a045ea6a541a7d6ba2aaca6db3dfad

**Comment by jerrywell** [comment]
@valerybodak there is barely difference between **ListView** and **SingleChildScrollView**. Based on [SingleChildScrollView's example](https://api.flutter.dev/flutter/widgets/SingleChildScrollView-class.html), all you need is putting **AutoScrollController** into the scroll view and wrap **AutoScrollTag** for your section inside the **Column** widget.

**Comment by lifetin** [comment]
必须得提供这个方法呀

我点击左边,需要跳到对应右边的位置上

![55](https://user-images.githubusercontent.com/10474968/56846927-ad9dd380-6907-11e9-90a0-fbd2aaacf098.png)

**Comment by tarobins** [comment]
here's a quick way to create a list that builds with the list scrolled to a particular index 

```
import 'package:flutter/material.dart';

final topKey = ValueKey('top');

class SelectedItemList extends StatelessWidget {
  final List<Widget> items;
  final int initialTopItemIndex;

  SelectedItemList({@required this.items, @required this.initialTopItemIndex});

  @override
  Widget build(BuildContext context) =>
      CustomScrollView(center: topKey, slivers: [
        SliverList(
          delegate: SliverChildListDelegate(
            items.sublist(0, initialTopItemIndex).reversed.toList(),
          ),
        ),
        SliverList(
          key: topKey,
          delegate: SliverChildListDelegate(
            items.sublist(initialTopItemIndex, initialTopItemIndex + 1),
          ),
        ),
        SliverList(
          delegate: SliverChildListDelegate(
            items.sublist(initialTopItemIndex + 1),
          ),
        ),
      ]);
}
```

**Comment by oseiasmribeiro** [comment]
> @oseiasmribeiro de bom grado! 
> Aqui está uma essência que criei com minhas amostras de código e descrições breves do que fiz. 
> Dê uma olhada e talvez discuta mais detalhes sobre a implementação. 
> https://gist.github.com/wtoalabi/51a045ea6a541a7d6ba2aaca6db3dfad

worked fine for me using ListView and expecting a little Delay. Wtoalabi Thank you! Does it work with SliverList?

**Comment by lifenautjoe** [comment]
Nested scroll views are such a pain in Flutter... The scroll-to-index package from quire-io works great with simple ListViews, as soon as there's silvers involved, the whole thing breaks down.

With 111 upvotes is this still part of the "Goals" milestone?

> P2: These are tasks we think are worth fixing in the coming years. It includes issues we've identified that might block fully shipping broad consumer-facing apps, correctness issues, and bugs relating to polish and quality. The date on this milestone is utterly arbitrary and just intended to sort the milestone appropriately.

**Comment by eseidelGoogle** [comment]
@goderbauer is the right person to comment on proper scheduling.

**Comment by mzimbres** [comment]
Please, implement this.

**Comment by pedromorgan** [comment]
> @Hixie is this a reasonable feature request?

How bizzare that `scrollToIndex` is still not implemented, whilst in the native.api's !!

**Comment by zhahao** [comment]
I have written a libary, see my [homepage](https://github.com/zhahao/list_view_item_builder),it provide like iOS UITableView, provide jumpTo(section,index) and animateTo(section,index,animation) functions.

**Comment by tomkastek** [comment]
I am having the same issues right now. Another idea I had is to add functionality to change if keepScrollOffset in a ScrollController should keep the offset to the end or the start of the list instead of start only so that on creation we are in the correct position already (also jumpTo(index) will show the item at index at top of the list. If we keep the offset we would be able to show the new items and the lowest new item would be viewable on the screen so that the user knows more data was loaded by eye and not by testing).

I also tried @tarobins solution as workaround. But it crashes if the top sliver is a CupertinoSliverRefreshControl.

**Comment by tomkastek** [comment]
@tarobins I get the following error output: https://gist.github.com/tomkastek/b63f75adfc735661391d84641764e73c

**Comment by tarobins** [comment]
Just for kicks, I commented out the assert that was causing the problem, but the CupertinoSliverRefreshControl still didn't work when its growthDirection was backwards.  

https://github.com/flutter/flutter/issues/38492

**Comment by slightfoot** [comment]
@marcglasberg your welcome 😃

**Comment by uwemeier-sclable** [comment]
@marcglasberg This does NOT work, if you have to use Slivers (inside a CustomScrollView), right?

**Comment by marcglasberg** [comment]
@uwemeier-sclable The idea itself works, but I would have to make a different version for CustomScrollView/Slivers, and it's non trivial.

**Comment by charleshan** [comment]
@uwemeier-sclable I think this might be what you're looking for. I found a bug while using this method though (#39715 ). Let me know this is the solution you're looking for and if you happened to find a solution for the bug 😁

> here's a quick way to create a list that builds with the list scrolled to a particular index
> 
> ```
> import 'package:flutter/material.dart';
> 
> final topKey = ValueKey('top');
> 
> class SelectedItemList extends StatelessWidget {
>   final List<Widget> items;
>   final int initialTopItemIndex;
> 
>   SelectedItemList({@required this.items, @required this.initialTopItemIndex});
> 
>   @override
>   Widget build(BuildContext context) =>
>       CustomScrollView(center: topKey, slivers: [
>         SliverList(
>           delegate: SliverChildListDelegate(
>             items.sublist(0, initialTopItemIndex).reversed.toList(),
>           ),
>         ),
>         SliverList(
>           key: topKey,
>           delegate: SliverChildListDelegate(
>             items.sublist(initialTopItemIndex, initialTopItemIndex + 1),
>           ),
>         ),
>         SliverList(
>           delegate: SliverChildListDelegate(
>             items.sublist(initialTopItemIndex + 1),
>           ),
>         ),
>       ]);
> }
> ```
> 
> If you need the item you're interested in at a position other than the top of the view port, you can use CustomScrollView's anchor property.

**Comment by tarobins** [comment]
Thanks @marcglasberg !  I've been slowly working on a solution as well.  

I've created a new widget, `ScrollablePositionedList`, that can be found [here](https://github.com/google/flutter.widgets/tree/master/lib/src/scrollable_positioned_list).  

The ScrollablePositionedList allows one to set an initial item, specified with `initialScrollIndex` to be positioned at the top of the viewport, or at another location in the viewport by using `initialAlignment`.  The ScrollablePositionedList also takes an `itemScrollController` that has  `jumpTo` and `scrollTo` methods.  The latter allows one to do an animated scroll to an arbitrary item in the list over some given duration and curve.  

The ScrollablePositionedList also accepts a `ItemPositionsListener` that tracks what items are currently visible as the user scrolls.

The list can be tried by using the gallery app included in the same package as ScrollablePositionedList and found [here](https://github.com/google/flutter.widgets/tree/master/gallery).  The currently last example in the gallery demos the ScrollablePositionedList.

A video of the list being scrolled can be found [here](https://drive.google.com/open?id=11UTjHc0R9WjFN8ipS2Z1HGopVWTK8ObY) and of some jumping and scrolling [here](https://drive.google.com/open?id=1X_Xe718XwYUr-1nxSQexIgmRdJVvuv_b).  Additionally, this [video](https://drive.google.com/open?id=1hBLfxtciTdJzFp7i2oNCliUleuf4duga) shows the first and last item fields at the bottom of the screen tracking the visible items as the user scrolls.

The ScrollablePositionedList is also available in [this dev release of the flutter.widgets library on pub](https://pub.dev/packages/flutter_widgets/versions/0.1.6-dev.1).  The widget required flutter 1.9.1, because that version of Flutter has some necessary bug fixes, so one must use the master channel of flutter until that version of Flutter makes its way to the other channels.  

ScrollablePositionedList is an early version. Please let me know what additional features would be useful here, or post bugs and feature requested [here](https://github.com/google/flutter.widgets/issues) in flutter.widgets' issue page.  The widget has some shortcomings, and I'd appreciate feedback on what the most important things to work on are.

**Comment by marcglasberg** [comment]
@tarobins That's great, but is there a reason you are not publishing each widget in a separate pub package? Are all these widgets connected and cannot work without one another? You maybe could publish them separately, it would help their use by the community.

Let me ask you a few questions about your list:
1) Only finite lists? Only infinite lists? Both finite and infinite lists?
2) Can you jump to any index instantly, or does it need to search and download/display all items (while scrolling from start-index to end-index)?

**Comment by tarobins** [comment]
I'll have to get back to you on the separate library issue.

1.  At the moment only finite lists.  I hope to remove that restriction at some point.  I'm not sure how hard that will be.

2. It can jump instantly.   It can also scroll 'instantly' in that it doesn't layout every item for long scrolss.   Obviously, it lays out enough items for the scroll effect and some extras to do scrolling smoothly for shorter scrolls.

**Comment by oseiasmribeiro** [comment]
@tarobins, sorry for my innocent question. Can I use the SliverPersistentHeader type along with ScrollablePositionList?

**Comment by tarobins** [comment]
@oseiasmribeiro 

No, at the moment, ScrollablePositionedList is basically the equivalent of ListView but with positioning based on items rather than pixel position.

ScrollablePositionedList is roughly based on the solution from [here](https://github.com/flutter/flutter/issues/12319#issuecomment-488170337) and at the moment, SliverPersistentHeader won't work properly above the centered sliver.  Beyond that, ScrollablePositionedList, actually is two list that are faded between to give the illusion of scrolling long distances (I can explain more if anyone wants), and I suspect we'd need to do some fidgeting to get floating slivered to appear as expected as we fade between lists.

It's possible that NestedScrollView could be used to solve this problem, but  NestedScrollView is complex, and someone would need to work on that for a bit.

**Comment by oseiasmribeiro** [comment]
Ok @tarobins. Thanks!

**Comment by marcglasberg** [comment]
@tarobins Seems you're on the path of finally solving this. If you want/accept contributions I'd be glad to help with the infinite part and documentation. Then if you publish it as a standalone library I can deprecate my package, and I believe this long standing issue can be closed once and for all.

**Comment by tarobins** [comment]
@marcglasberg  will send you email.

**Comment by rockingdice** [comment]
@tarobins Thanks for your great work! 
Are the other libraries are necessary to use the list? I'm planning to move the codes into my project.

**Comment by tarobins** [comment]
@marcglasberg and anyone else interested, we can accept PRs.  When you submit one, and once the review is done, I'll need to manually merge your change into Google's internal code repo and then push it back to flutter.widgets.  

The widget will have to stay in the repo it's in for the time being, but as @rockingdice confirmed, the other widgets in that repo are not needed.

@rockingdice  thanks for posting the issue.

**Comment by droidluv** [comment]
@tarobins Yup would love if [this](https://github.com/google/flutter.widgets/issues/31) was fixed, right now using trailing edge for initialAlignment is causing an exception when its outside the asserted range, though just working with itemPositionListener and itemScrollIndex is more than what I could've asked for, thanks for the highly useful widget though, it can be used to enhance user experience greatly when working with modals and lists.

So just in case does this have any special caveats like performance constraints or some gotcha's which I have to be careful about? I might be cribbing at this point now but can you also add a separated builder method as well like the ListView (though not really a concern)

lol @ the annoyance tag, Is there any particular reason why this widget is not a part of the framework or adopt some of its capabilities, seems like a pretty standard feature to me.

**Comment by patelkrunal78** [comment]
If you use a shrink-wrapped ListView with reverse: true, scrolling it to 0.0 will do what you want.

Full Code At [Here](https://icetutor.com/question/programmatically-scrolling-to-the-end-of-a-listview-2)

**Comment by sooxiaotong** [comment]
@lifetin 抱歉离题一下，能教我怎么实现你截图的那个界面吗？

**Comment by tarobins** [comment]
@droidluv  
I've submitted a fix for https://github.com/google/flutter.widgets/issues/31 to the internal google review system.  Will take a bit to filter to public repo.

**Comment by gaaclarke** [comment]
Wow, there's been a lot of great discussion around this.  I'll try to resolve this issue.  I'll first show you how to accomplish what you want, then explain why the issue is ambiguous, then explain what we might change in the Flutter framework.

**The Solution**
Many people have suggested this in one form or another but the gist is you want to do the following things:
1) Create an object that represents the data model for your ListView.
1) Make your ListView build its contents based on that data model.
1) Query the data model to find the offset of the item you are interested in.

Here is a helper interface that accomplishes this:
```dart
/// An interface for representing the contents of a ListView.
class ListViewDataSource {
  /// The number of items that will be displayed in the ListView.
  int get numItems => 0;

  /// The height of the widget at [index].
  double itemHeight(int index) => 0;

  /// Generate the widget at [index].
  Widget build(BuildContext context, int index) => null;

  /// Calculates the offset of a widget at [index].  If you scroll to that
  /// index, that item will be at the top of the ListView.
  static double itemOffset(ListViewDataSource dataSource, int index) {
    if (index >= dataSource.numItems) {
      throw new RangeError("index out of range");
    }
    double result = 0.0;
    for (int i = 0; i < index; ++i) {
      result += dataSource.itemHeight(i);
    }
    return result;
  }

  /// A factory function that makes a [ListView] that displays the contents of a
  /// [ListViewDataSource].
  static ListView makeListView(
      {ScrollController controller,
      EdgeInsetsGeometry padding,
      @required ListViewDataSource dataSource}) {
    return ListView.builder(
      controller: controller,
      padding: padding,
      itemCount: dataSource.numItems,
      itemBuilder: dataSource.build,
    );
  }
}
```
You'll find a gist that shows how to use this with ScrollerController [here](https://gist.github.com/gaaclarke/4200741c4413707f0f73a887bf5d9b39).

**Why the problem is ambiguous**
When you say you want to scroll to an item at index, do you want to scroll so that that item is at the top of the ListView?  or do you want to scroll so that it is in the center or bottom of the ListView?  Maybe something slightly different?

The better API would give you the offset of an item at index which would allow the user to decide exactly how they want to scroll to that item.

**Gaaclarke's recommendation**
Now the question about what, if anything, we should add to the Flutter API.

If I had been around for when ListView had been designed, I would have suggested the `ListView.builder` API matched more what I showed in the solution section, one that takes in a data model instead of a builder function.  We could add yet another factory function that takes in a data model.  I have concerns that that would confuse users by having two similar ways of creating a ListView, and we aren't going to remove `ListView.builder`.

So, my recommendation is that we don't change or add anything to the Flutter API.  Instead we should update the [documentation on ListView](https://api.flutter.dev/flutter/widgets/ListView-class.html) to include at a minimum a usage of ScrollerController.  Maybe we could add an example of using a data source as well, I feel less strongly about that.

**Comment by pedromorgan** [comment]
All it needs to do is "select" and item in the select and make it visible if its off screen. Simple as that. The workaround etc seem a mess

**Comment by Hixie** [comment]
@gaaclarke The problems with that approach are that implementing `itemHeight` is essentially impossible, and `index` might be very large (such that walking through all the items from zero to index is prohibitively expensive), and the viewport might contain more than just a sliver list, it could contain a grid plus a list plus some headings plus an appbar plus another list.

A more efficient implementation for the simple case of a single ListView is something where you split the list in two at the point you want to jump to, put the first list at negative indices and the second list at positive indices, and jump to zero. However, that fails a major use case which is animating to the index, and of course doesn't handle more subtle cases like mixed viewports as described above.

**Comment by gaaclarke** [comment]
@Hixie thanks for the response.

> implementing `itemHeight` is essentially impossible (edit: or at least prohibitively expensive)

Why?  I did it just fine in the sample code.  If you have a data model, it is the definition of itemHeight, no calculation needed.

> viewport might contain more than just a sliver list

That should be represented in your data model so you can query it as well.

> A more efficient implementation for the simple case

Why is that more efficient?  What is the inefficiency in the proposed data model approach?

**Comment by tarobins** [comment]
> @Hixie thanks for the response.
> 
> > implementing `itemHeight` is essentially impossible (edit: or at least prohibitively expensive)
> 
> Why? I did it just fine in the sample code. If you have a data model, it is the definition of itemHeight, no calculation needed.

In some cases you can't know what the height of an item is until you perform a layout of the item and take into account issues such as the space used for a particular font, line breaking of text, and the size of images, the latter of which might need to be downloaded from some remote source.  Having your data source return the height of a item would mean the data source will need to do this layout (and would consequently need to know something about the dimensions of the viewport, such as needing to know the width of the viewport to work out line breaking).  In the case of a long list, you could need to do the layout for thousands of items to compute the position of some items, leading to the expensive computations @Hixie refers to.

**Comment by gaaclarke** [comment]
@tarobins That would only be the case if you are display content that isn't known at build time and is in a form factor that isn't defined at build time and can't be spatially compressed.  For example, showing user submitted images but you refuse to display them resized.

Even in your text situation, the height of a widget is deterministic and should be able to be calculated.  I don't know if we've surfaced the logic to the framework but there is definitely code in the engine that given a string and a font and a width, it will tell you the height of the space needed to render it completely.  It isn't the slickest solution, but I don't think the problem is all that common either.  A chat room app might be the only thing I could think of.   Most people can determine the height of their widgets or design their widgets to render in a specified height.

I know this because the original post asks why we can't do something that UITableView does, and that's because [UITableView has a data model](https://developer.apple.com/documentation/uikit/uitableviewdelegate/1614998-tableview) that looks similar to what I'm proposing.  Without a data model, I dare say it's impossible to efficiently jump to a given item in a situation with heterogeneous children heights.  The only solution there is to build and throw away all the proceeding widgets, which is almost in all cases going to be unacceptable.

**Comment by tarobins** [comment]
> Without a data model, I dare say it's impossible to efficiently jump to a given item in a situation with heterogeneous children heights. The only solution there is to build and throw away all the proceeding widgets, which is almost in all cases going to be unacceptable.

I'm not quite sure exactly what you mean by data model in this context.  Do you mean a model that provides heights for the items or just a model of the data to be displayed?

We've solved the jumping case without knowing the heights of the items in a data model and without building all the widgets by using the method @Hixie described.  And example can be found here https://github.com/flutter/flutter/issues/12319#issuecomment-488170337

Smooth scrolling is a bit more complicated.  There's one solution here https://github.com/google/flutter.widgets/tree/master/lib/src/scrollable_positioned_list

**Comment by gaaclarke** [comment]
> I'm not quite sure exactly what you mean by data model in this context. Do you mean a model that provides heights for the items or just a model of the data to be displayed?

For a ListView you want to display some state `s[n]`.  `s[n]` is the data model.  There also has to be a function `f(s[x]) -> itemHeight`.  I used data model to mean `s` and `f`.

> We've solved the jumping case without knowing the heights of the items in a data model and without building all the widgets by using the method @Hixie described. And example can be found here [#12319 (comment)](https://github.com/flutter/flutter/issues/12319#issuecomment-488170337)

That's an interesting approach but I think a solution that doesn't involve ScrollerController is kind of hacky in my opinion.  Controlling where you are scrolled is raison-d-etre of ScrollerController.

> Smooth scrolling is a bit more complicated. There's one solution here https://github.com/google/flutter.widgets/tree/master/lib/src/scrollable_positioned_list

It looks like this is caching the size of widgets, introspecting into the offset as widgets are created?  It's a lot of code to accomplish something similar to what I did in 100 lines of code, just to avoid defining `f(s[x])`.  It would not perform as well in non-scrolling case but I can see how it is easier to use for some data.  I'll give it a more thorough read tomorrow.

**Comment by tarobins** [comment]
> That's an interesting approach but I think a solution that doesn't involve ScrollerController is kind of hacky in my opinion. Controlling where you are scrolled is raison-d-etre of ScrollerController.

Even if you know what  pixel offset to start the ScrollController at, or jump to, the current implementation of ListView will build and layout everything between the 0 offset and your desired offset in order to know what to put at the scroll offset your asked for, so, for long lists, using the ScrollController to position the list can be inefficient.

**Comment by knuesel** [comment]
@gaaclarke there are many cases where the ListView items are generated programmatically in such a way that the dimensions cannot be known statically. For example I'm working on an application that shows a list of tasks in a ListView. Each task has dynamic elements (mainly text widgets) that the user can configure, so I rely on Flutter's layout engine to determine the dimensions of the ListView items. I'm using a horizontal list view, where dynamic dimensions along the main axis are probably more common, but the same problem would arise for a vertical list of user-editable tasks.

**Comment by tarobins** [comment]
https://github.com/google/flutter.widgets/issues/31
and https://github.com/google/flutter.widgets/issues/23 should be fixed now.

**Comment by oseiasmribeiro** [comment]
Tarobins, ScrollablePositionedList worked beautifully. Help me now add a header that appears when scrolling down and disappears when scrolling up. This is very useful for showing details of a certain book, as in the Bible.

**Comment by tarobins** [comment]
Yeah, SliverPersistentHeader doesn't work with ScrollablePositionedList nor
the method from
https://github.com/flutter/flutter/issues/12319#issuecomment-488170337.
It's possible SliverPersistentHeader might work with one of those when used
within NestedScrollView, but I haven't tried yet.  I'll give it a try
today, but no promises I'll get anywhere.

https://github.com/google/flutter.widgets/issues/32 is the relevant issue,
though a subset solution of that issue might be more tractable in the short
term.

**Comment by oseiasmribeiro** [comment]
Thanks a lot already Tarobins. ScrollablePositionedList is working very well. Congratulations!

**Comment by oseiasmribeiro** [comment]
![Captura de Tela 2019-11-29 às 14 31 21](https://user-images.githubusercontent.com/36827173/69884040-628cda00-12b5-11ea-9be4-35c724d85f45.png)

Tarobins, with NestedScrollView I have managed to add the header, but Pinned and Floating don't seem to work. I can't make the Header it disappear by rolling up and appear by rolling down.

**Comment by tarobins** [comment]
I'm not super surprised.  I'm just creating my experiment app, so I'll play around with it for a bit and see what's going on.

**Comment by oseiasmribeiro** [comment]
Ok thanks.

**Comment by tarobins** [comment]
So, first, I noticed that in general floating doesn't seem to work as I expected for NestedScrollView.  I tried setting in the example from https://api.flutter.dev/flutter/widgets/NestedScrollView-class.html and the SliverAppBar didn't float as shown in SliverAppBar's docs.

I didn't have high hopes to ScrollablePositionedList to work out of the box within a NestedScrollView, and in particular because in the example for NestedScrollView it is noted for its CustomScrollView
```
// The "controller" and "primary" members should be left
// unset, so that the NestedScrollView can control this
// inner scroll view.
```
and that's not true of ScrollablePositionedList's internal CustomScrollView.  Furthermore, ScrollablePositionedList does some magic with switching what CustomScrollView is displayed sometimes, which is likely to confuse NestedScrollView.

Some work will be needed to support SliverAppBar within ScrollablePositionedList

However, if all you care about is positioning an item at the top of the list when you display it, this worked surprisingly well:

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

final topKey = ValueKey('top');

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Material(
        child: SafeArea(
          child: body,
        ),
      ),
    );
  }
}

Widget get body => NestedScrollView(
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return <Widget>[
          SliverAppBar(
            title: const Text('Books'),
            // This is the title in the app bar.
            pinned: true,
            expandedHeight: 200.0,
            forceElevated: true,
          ),
        ];
      },
      body: SafeArea(
        top: false,
        bottom: false,
        child: CustomScrollView(
          center: topKey,
          slivers: <Widget>[
            SliverFixedExtentList(
              itemExtent: 48.0,
              delegate: SliverChildBuilderDelegate(
                (BuildContext context, int index) {
                  return ListTile(
                    title: Text('Item ${14 - index}'),
                  );
                },
                childCount: 15,
              ),
            ),
            SliverFixedExtentList(
              key: topKey,
              itemExtent: 48.0,
              delegate: SliverChildBuilderDelegate(
                (BuildContext context, int index) {
                  return ListTile(
                    title: Text('Item 15'),
                  );
                },
                childCount: 1,
              ),
            ),
            SliverFixedExtentList(
              itemExtent: 48.0,
              delegate: SliverChildBuilderDelegate(
                (BuildContext context, int index) {
                  return ListTile(
                    title: Text('Item ${index + 16}'),
                  );
                },
                childCount: 15,
              ),
            ),
          ],
        ),
      ),
    );
```

**Comment by oseiasmribeiro** [comment]
Great! Thanks. This way I how get the index of scroll? I save him to back read the book where he left off.

**Comment by tarobins** [comment]
oh. that's can't be done with the above.  hmmm.  

If the size of your content and screen never changes, you could just store the scroll offset https://api.flutter.dev/flutter/widgets/ScrollController/offset.html

If you can do without a collapsing app bar, then you can just have a statically sized widget above the ScrollablePositionedList with your header.  Then use the ScrollablePositionedList's itemPositionsListener to get the position.

otherwise, it's trickier.  

I really should try to support a SliverAppBar on ScrollablePositionedList.  I've had that request a couple times.  I'll try to find some time to work on it.  There's a few non-obvious behaviours we need to spec out though, such as does using ItemScrollController.scrollTo and ItemScrollController.jumpTo affect the size of the app bar and how so.

**Comment by oseiasmribeiro** [comment]
Ok. For now I go use Fixed Header, until you implement something similar in ScrollablePositionedList. Thank you so much for everything.

**Comment by tarobins** [comment]
I'm going to close this issue with ScrollablePositionedList https://github.com/google/flutter.widgets/tree/master/lib/src/scrollable_positioned_list available.  I am still working on the library and there are some fixes already submitted to the internal google repo that should be pushed to the github repo soon.  Please file additional bugs and feature requests here: https://github.com/google/flutter.widgets/issues

**Comment by duzenko** [comment]
Why is it closed with an external link? Are you not going to fix this with standard ListView at all?

**Comment by tarobins** [comment]
I don't see a solution that can be used on ListView itself without making a mess of the API because there's a conflict between pixel-based and item-based scrolling.  When we jump or scroll to a item that's far down a list, if we want to keep the pixel-based and item-based indices consistent, we'd need to layout every item before the desired item, which can be prohibitively expensive.

**Comment by duzenko** [comment]
Please let us decide if we can live with the 'expenses'.
Otherwise consider backporting the widget you linked to above in flutter core.
Please do not close critical functionality holes with a no-fix.

**Comment by tarobins** [comment]
I can understand that one would want a solution integrated with the existing ListView; however, I wasn't able to think of one that I thought would be workable and at the same time that I thought would be an acceptable change to the core ListView. I did create a solution that I thought got to the core of what was originally asked for in this issue. 

I know there are still some desired features.  For example, I have seen the requests to have the ScrollablePositionedList work with a collapsible appbar, unfortunately I haven't had time to explore a solution to that use case.  (More requests can be seen or added at https://github.com/google/flutter.widgets/issues, and people are encouraged to fork my solution and submit PRs, or just create new solutions based on mine or from scratch).  If you have use cases not covered, please open a new issue (perhaps with a linking comment from here so people interested see the new issue), with an exact explanation of not-yet-covered needs and use cases.  Doing so will help someone provide solutions that meet the use cases you have.  Also, if more documentation is needed for the solution I provided, please let me know, and I’ll work on providing updates, as I do want people to be able to use it. 

As for this issue, given that I am the current assignee of the issue, I didn’t want to leave the issue open, and give the impression that I was still working on the issue past the solution I was able to come up with. However, if someone does want to continue to work on integrating a solution with ListView, I'm happy to hand over this issue and they can reopen it. However, if no one has a feasible change to ListView in mind, I'm not sure leaving this issue open is useful, as doing so might give people an incorrect impression that a change to ListView is being worked on, or eventually will be worked on at some point in the reasonable future.  If such a change is not foreseeable, it might be best to direct energies to coming up with new widgets that meet peoples’ use cases, which, personally, I think is the more obtainable path.

**Comment by marcglasberg** [comment]
@tarobins However, it would be great if this widget could have it's own package in pub/dev, instead of being bundled with other widgets. That's totally non-standard. I am guilty of doing that myself (https://pub.dev/packages/assorted_layout_widgets) but I think ScrollablePositionedList is just too important to be buried into a package called flutter.widgets.

**Comment by tarobins** [comment]
exciting news!  ScrollablePositionedList is now available in its own package on pub:  https://pub.dev/packages/scrollable_positioned_list

Please start pointing to that package.

**Comment by tarobins** [comment]
Thanks particularly to @mehmetf
